File: sdk/go/verifier.go
package iam

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/lestrrat-go/jwx/v2/jwk"
	"github.com/lestrrat-go/jwx/v2/jwt"
)

const defaultIssuer = "teamusers"

// Option configures any SDK client. Options that do not apply to a specific
// constructor are ignored by that constructor.
type Option func(any)

// VerifierOption is the component-specific name for Option.
type VerifierOption = Option

// PermissionsOption is the component-specific name for Option.
type PermissionsOption = Option

// ClientOption is the component-specific name for Option.
type ClientOption = Option

// WithHTTPClient supplies the HTTP client used for JWKS and permission calls.
func WithHTTPClient(client *http.Client) Option {
	return func(target any) {
		switch value := target.(type) {
		case *Verifier:
			if client != nil {
				value.httpClient = client
			}
		case *PermissionsClient:
			if client != nil {
				value.httpClient = client
			}
		}
	}
}

// WithIssuer overrides the expected JWT issuer. The service default is
// "teamusers"; the JWKS URL is independent from the issuer claim.
func WithIssuer(issuer string) Option {
	return func(target any) {
		if value, ok := target.(*Verifier); ok {
			value.issuer = strings.TrimSpace(issuer)
		}
	}
}

// WithExpectedIssuer is an alias for WithIssuer.
func WithExpectedIssuer(issuer string) Option {
	return WithIssuer(issuer)
}

// WithJWKSMinRefreshInterval controls the minimum background refresh interval
// used by the jwx key cache.
func WithJWKSMinRefreshInterval(interval time.Duration) Option {
	return func(target any) {
		if value, ok := target.(*Verifier); ok && interval > 0 {
			value.jwksMinRefresh = interval
		}
	}
}

// WithServiceToken configures the bearer token used for service-only APIs.
func WithServiceToken(token string) Option {
	return func(target any) {
		if value, ok := target.(*PermissionsClient); ok {
			value.serviceToken = strings.TrimSpace(token)
		}
	}
}

// WithTokenSource configures a callback that supplies a current service token.
// It takes precedence over a static token.
func WithTokenSource(source func() (string, error)) Option {
	return func(target any) {
		if value, ok := target.(*PermissionsClient); ok {
			value.tokenSource = source
		}
	}
}

// WithPermissionsHTTPClient is an explicit alias for WithHTTPClient.
func WithPermissionsHTTPClient(client *http.Client) Option {
	return WithHTTPClient(client)
}

// WithPermissionsTTL configures the in-process permission entry TTL.
func WithPermissionsTTL(ttl time.Duration) Option {
	return func(target any) {
		if value, ok := target.(*PermissionsClient); ok && ttl >= 0 {
			value.ttl = ttl
		}
	}
}

// WithTTL is an alias for WithPermissionsTTL.
func WithTTL(ttl time.Duration) Option {
	return WithPermissionsTTL(ttl)
}

// WithRemoteOnly makes Client.Allow use the remote authorization endpoint
// rather than the local permission cache.
func WithRemoteOnly(remoteOnly bool) Option {
	return func(target any) {
		if value, ok := target.(*Client); ok {
			value.RemoteOnly = remoteOnly
		}
	}
}

// Verifier verifies EdDSA access tokens against a cached JWKS document.
type Verifier struct {
	issuerBaseURL  string
	jwksURL        string
	issuer         string
	httpClient     *http.Client
	jwksMinRefresh time.Duration
	cache          *jwk.Cache
	cacheContext   context.Context
	cacheCancel    context.CancelFunc
	registerError  error
	closeOnce      sync.Once
}

// NewVerifier constructs a lazy JWKS verifier. Construction does not require
// the issuer to be reachable; Verify returns the fetch error until the cache
// has obtained a usable JWKS document.
func NewVerifier(issuerBaseURL string, opts ...VerifierOption) *Verifier {
	base, jwksURL, err := normalizeIssuerURL(issuerBaseURL)
	cacheContext, cancel := context.WithCancel(context.Background())
	verifier := &Verifier{
		issuerBaseURL:  base,
		jwksURL:        jwksURL,
		issuer:         defaultIssuer,
		httpClient:     http.DefaultClient,
		jwksMinRefresh: time.Hour,
		cacheContext:   cacheContext,
		cacheCancel:    cancel,
		registerError:  err,
	}
	for _, opt := range opts {
		if opt != nil {
			opt(verifier)
		}
	}
	if verifier.issuer == "" {
		verifier.issuer = defaultIssuer
	}
	if err == nil {
		verifier.cache = jwk.NewCache(cacheContext)
		registerOpts := []jwk.RegisterOption{jwk.WithMinRefreshInterval(verifier.jwksMinRefresh)}
		if verifier.httpClient != nil {
			registerOpts = append(registerOpts, jwk.WithHTTPClient(verifier.httpClient))
		}
		if registerErr := verifier.cache.Register(jwksURL, registerOpts...); registerErr != nil {
			verifier.registerError = registerErr
		}
	}
	return verifier
}

// Close stops the jwx cache's background workers. It is safe to call more than
// once and is optional for process-lifetime verifiers.
func (v *Verifier) Close() error {
	if v == nil {
		return nil
	}
	v.closeOnce.Do(func() {
		if v.cacheCancel != nil {
			v.cacheCancel()
		}
	})
	return nil
}

// Verify validates an access token, including EdDSA signature, issuer,
// expiration, subject, kind, and permission version claims.
func (v *Verifier) Verify(ctx context.Context, raw string) (Claims, error) {
	if v == nil {
		return Claims{}, errors.New("nil verifier")
	}
	if ctx == nil {
		ctx = context.Background()
	}
	if v.registerError != nil {
		return Claims{}, fmt.Errorf("configure JWKS cache: %w", v.registerError)
	}
	if strings.TrimSpace(raw) == "" {
		return Claims{}, errors.New("access token is empty")
	}
	if v.cache == nil || v.jwksURL == "" {
		return Claims{}, errors.New("JWKS cache is unavailable")
	}
	set, err := v.cache.Get(ctx, v.jwksURL)
	if err != nil {
		return Claims{}, fmt.Errorf("fetch JWKS: %w", err)
	}

	if kid, ok := tokenKeyID(raw); ok {
		if _, present := set.LookupKeyID(kid); !present {
			refreshed, refreshErr := v.cache.Refresh(ctx, v.jwksURL)
			if refreshErr != nil {
				return Claims{}, fmt.Errorf("refresh JWKS for key %q: %w", kid, refreshErr)
			}
			set = refreshed
		}
	}

	token, err := jwt.Parse([]byte(raw), jwt.WithKeySet(set), jwt.WithValidate(true), jwt.WithIssuer(v.issuer))
	if err != nil {
		return Claims{}, fmt.Errorf("verify access token: %w", err)
	}
	if token.Issuer() != v.issuer {
		return Claims{}, errors.New("invalid access token issuer")
	}
	expiry := token.Expiration()
	if expiry.IsZero() || !expiry.After(time.Now()) {
		return Claims{}, errors.New("access token is expired")
	}
	subject := token.Subject()
	if subject == "" {
		return Claims{}, errors.New("access token subject is missing")
	}
	kind, ok := stringClaim(token, "kind")
	if !ok || (kind != "user" && kind != "service") {
		return Claims{}, errors.New("invalid access token kind")
	}
	permVer, ok := int64Claim(token, "perm_ver")
	if !ok || permVer < 0 {
		return Claims{}, errors.New("invalid access token perm_ver")
	}
	team := ""
	if value, present := token.Get("team"); present {
		var teamOK bool
		team, teamOK = value.(string)
		if !teamOK {
			return Claims{}, errors.New("invalid access token team")
		}
	}
	return Claims{Subject: subject, Team: team, Kind: kind, PermVer: permVer, Expiry: expiry}, nil
}

func normalizeIssuerURL(raw string) (string, string, error) {
	base := strings.TrimRight(strings.TrimSpace(raw), "/")
	if base == "" {
		return "", "", errors.New("issuer base URL is empty")
	}
	parsed, err := url.Parse(base)
	if err != nil || parsed.Scheme == "" || parsed.Host == "" {
		if err == nil {
			err = errors.New("URL must include scheme and host")
		}
		return base, "", fmt.Errorf("invalid issuer base URL: %w", err)
	}
	return base, base + "/.well-known/jwks.json", nil
}

func tokenKeyID(raw string) (string, bool) {
	parts := strings.Split(raw, ".")
	if len(parts) != 3 {
		return "", false
	}
	header, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return "", false
	}
	var values struct {
		KeyID string `json:"kid"`
	}
	if err := json.Unmarshal(header, &values); err != nil || values.KeyID == "" {
		return "", false
	}
	return values.KeyID, true
}

func stringClaim(token jwt.Token, name string) (string, bool) {
	value, ok := token.Get(name)
	if !ok {
		return "", false
	}
	text, ok := value.(string)
	return text, ok
}

func int64Claim(token jwt.Token, name string) (int64, bool) {
	value, ok := token.Get(name)
	if !ok {
		return 0, false
	}
	switch number := value.(type) {
	case int:
		return int64(number), true
	case int64:
		return number, true
	case int32:
		return int64(number), true
	case uint:
		return int64(number), uint64(number) <= uint64(^uint64(0)>>1)
	case uint64:
		return int64(number), number <= uint64(^uint64(0)>>1)
	case float64:
		return int64(number), number >= 0 && number == float64(int64(number))
	case json.Number:
		parsed, err := number.Int64()
		return parsed, err == nil
	default:
		return 0, false
	}
}

File: sdk/ts/src/verifier.ts
import { createLocalJWKSet, jwtVerify } from "jose";
import type { JSONWebKeySet as JoseJSONWebKeySet, LocalJWKSet } from "jose";

const DEFAULT_ISSUER = "teamusers";
const DEFAULT_AUDIENCE = "teamusers";
const DEFAULT_CACHE_TTL_MS = 60 * 60 * 1000;

type JoseJWKS = JoseJSONWebKeySet;

/** A JSON response with the subset needed by the verifier's JWKS loader. */
export interface JWKSResponse {
  readonly ok: boolean;
  readonly status: number;
  json(): Promise<unknown>;
}

/** Injectable fetch function, useful for tests and custom HTTP transports. */
export type JWKSFetcher = (url: string) => Promise<JWKSResponse>;

export interface JSONWebKeySet {
  readonly keys: readonly Record<string, unknown>[];
}

export interface VerifierOptions {
  /** Expected JWT audience. Defaults to `teamusers`. */
  readonly audience?: string;
  /** Alias for `audience`. */
  readonly expectedAudience?: string;
  /** JWKS cache lifetime in milliseconds. Defaults to one hour. */
  readonly cacheTtlMs?: number;
  /** Alias for `cacheTtlMs`. */
  readonly ttlMs?: number;
  /** Custom JWKS transport. Defaults to global `fetch`. */
  readonly fetch?: JWKSFetcher;
  /** Alias for `fetch`. */
  readonly fetcher?: JWKSFetcher;
  /** Optional in-memory JWKS, intended for deterministic tests. */
  readonly jwks?: JSONWebKeySet;
}

export type TokenKind = "user" | "service";
export type Audience = string | readonly string[];

/** Identity claims carried by a successfully verified access token. */
export class Claims {
  public readonly subject: string;
  public readonly team: string;
  public readonly kind: TokenKind;
  public readonly permVer: number;
  public readonly expiry: Date;
  public readonly audience: Audience;

  public constructor(values: {
    readonly subject: string;
    readonly team: string;
    readonly kind: TokenKind;
    readonly permVer: number;
    readonly expiry: Date;
    readonly audience: Audience;
  }) {
    this.subject = values.subject;
    this.team = values.team;
    this.kind = values.kind;
    this.permVer = values.permVer;
    this.expiry = values.expiry;
    this.audience = values.audience;
  }

  /** JWT `sub` spelling for callers working directly with token claims. */
  public get sub(): string {
    return this.subject;
  }

  /** JWT `perm_ver` spelling for callers working directly with token claims. */
  public get perm_ver(): number {
    return this.permVer;
  }

  /** JWT expiration as a Unix timestamp in seconds. */
  public get exp(): number {
    return this.expiry.getTime() / 1000;
  }

  /** JWT `aud` spelling for callers working directly with token claims. */
  public get aud(): Audience {
    return this.audience;
  }
}

export type ErrorCode =
  | "INVALID_VERIFIER_CONFIGURATION"
  | "JWKS_FETCH_FAILED"
  | "TOKEN_EMPTY"
  | "TOKEN_VERIFICATION_FAILED"
  | "TOKEN_CLAIMS_INVALID"
  | "UNAUTHORIZED"
  | "FORBIDDEN"
  | "PERMISSION_FETCH_FAILED"
  | "AUTHORIZATION_FAILED"
  | "NATS_UNAVAILABLE";

/** Base class for all errors raised by this SDK. */
export class SDKError extends Error {
  public readonly code: ErrorCode;

  public constructor(code: ErrorCode, message: string, cause?: unknown) {
    super(message, cause === undefined ? undefined : { cause });
    this.name = new.target.name;
    this.code = code;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

/** The verifier could not be configured with a usable base URL or options. */
export class VerifierConfigError extends SDKError {
  public constructor(message: string, cause?: unknown) {
    super("INVALID_VERIFIER_CONFIGURATION", message, cause);
  }
}

/** The JWKS endpoint did not return a usable key set. */
export class JWKSFetchError extends SDKError {
  public constructor(message: string, cause?: unknown) {
    super("JWKS_FETCH_FAILED", message, cause);
  }
}

/** The token could not be authenticated with EdDSA or its JOSE claims failed. */
export class TokenVerificationError extends SDKError {
  public constructor(message: string, cause?: unknown) {
    super("TOKEN_VERIFICATION_FAILED", message, cause);
  }
}

/** A cryptographically valid token carried invalid or incomplete application claims. */
export class TokenClaimsError extends TokenVerificationError {
  public readonly code = "TOKEN_CLAIMS_INVALID" as const;

  public constructor(message: string, cause?: unknown) {
    super(message, cause);
    this.name = new.target.name;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

// Descriptive aliases for applications that use the shorter error names.
export { SDKError as TeamusersError };
export { TokenVerificationError as InvalidTokenError };
export { TokenClaimsError as InvalidClaimsError };

interface CacheEntry {
  readonly keySet: LocalJWKSet;
  readonly expiresAt: number;
}

/**
 * Verifies teamusers Ed25519 access tokens against a cached JWKS document.
 *
 * Construction is lazy: the JWKS endpoint is not contacted until `verify` is
 * called. A custom fetcher or in-memory key set can be supplied for tests.
 */
export class Verifier {
  private readonly jwksURL: string;
  private readonly expectedAudience: string;
  private readonly cacheTtlMs: number;
  private readonly fetcher: JWKSFetcher;
  private readonly staticKeySet?: LocalJWKSet;
  private readonly configError?: VerifierConfigError;
  private cache?: CacheEntry;
  private inFlightKeySet?: Promise<LocalJWKSet>;

  public constructor(baseURL: string, options?: VerifierOptions | string) {
    const resolvedOptions: VerifierOptions =
      typeof options === "string" ? { audience: options } : options ?? {};
    const base = baseURL.trim().replace(/\/+$/, "");

    let parsedBase: URL | undefined;
    try {
      parsedBase = new URL(base);
      if (parsedBase.protocol === "" || parsedBase.host === "") {
        throw new Error("URL must include scheme and host");
      }
    } catch (error) {
      this.configError = new VerifierConfigError("invalid verifier base URL", error);
      this.jwksURL = "";
      this.expectedAudience = DEFAULT_AUDIENCE;
      this.cacheTtlMs = DEFAULT_CACHE_TTL_MS;
      this.fetcher = async () => {
        throw this.configError;
      };
      return;
    }

    const configuredAudience =
      resolvedOptions.audience ?? resolvedOptions.expectedAudience ?? DEFAULT_AUDIENCE;
    if (configuredAudience.trim() === "") {
      this.configError = new VerifierConfigError("expected audience is empty");
    }
    this.expectedAudience = configuredAudience;
    this.jwksURL = `${parsedBase.toString().replace(/\/$/, "")}/.well-known/jwks.json`;

    const configuredTTL = resolvedOptions.cacheTtlMs ?? resolvedOptions.ttlMs ?? DEFAULT_CACHE_TTL_MS;
    if (!Number.isFinite(configuredTTL) || configuredTTL < 0) {
      this.configError = new VerifierConfigError("JWKS cache TTL must be a finite non-negative number");
      this.cacheTtlMs = DEFAULT_CACHE_TTL_MS;
    } else {
      this.cacheTtlMs = configuredTTL;
    }

    this.fetcher =
      resolvedOptions.fetcher ??
      resolvedOptions.fetch ??
      (async (url: string) => globalThis.fetch(url));

    if (resolvedOptions.jwks !== undefined) {
      try {
        this.staticKeySet = createLocalJWKSet(toJoseJWKS(resolvedOptions.jwks));
      } catch (error) {
        this.configError = new VerifierConfigError("invalid in-memory JWKS", error);
      }
    }
  }

  /** Release verifier resources; retained for parity with the Go SDK. */
  public close(): void {}

  /** Verify a compact JWT and return its typed teamusers identity claims. */
  public async verify(raw: string): Promise<Claims> {
    if (this.configError !== undefined) {
      throw this.configError;
    }
    if (raw.trim() === "") {
      throw new TokenVerificationError("access token is empty");
    }

    let keySet = await this.getKeySet();
    let verified: Awaited<ReturnType<typeof jwtVerify>>;
    try {
      verified = await this.verifyWithKeySet(raw, keySet);
    } catch (error) {
      if (!isNoMatchingKeyError(error)) {
        throw new TokenVerificationError("verify access token", error);
      }
      keySet = await this.getKeySet(true);
      try {
        verified = await this.verifyWithKeySet(raw, keySet);
      } catch (retryError) {
        throw new TokenVerificationError("verify access token", retryError);
      }
    }

    return claimsFromPayload(verified.payload, this.expectedAudience);
  }

  private async verifyWithKeySet(raw: string, keySet: LocalJWKSet) {
    return jwtVerify(raw, keySet, {
      algorithms: ["EdDSA"],
      issuer: DEFAULT_ISSUER,
      audience: this.expectedAudience,
    });
  }

  private async getKeySet(forceRefresh = false): Promise<LocalJWKSet> {
    if (this.staticKeySet !== undefined) {
      return this.staticKeySet;
    }
    const now = Date.now();
    if (!forceRefresh && this.cache !== undefined && this.cache.expiresAt > now) {
      return this.cache.keySet;
    }

    // A single promise covers both ordinary expiry loads and forced key-miss
    // refreshes. This prevents a burst of verifications from stampeding JWKS.
    if (this.inFlightKeySet !== undefined) {
      return this.inFlightKeySet;
    }

    const fetchPromise = this.fetchAndCacheKeySet();
    this.inFlightKeySet = fetchPromise;
    try {
      return await fetchPromise;
    } finally {
      if (this.inFlightKeySet === fetchPromise) {
        this.inFlightKeySet = undefined;
      }
    }
  }

  private async fetchAndCacheKeySet(): Promise<LocalJWKSet> {
    let response: JWKSResponse;
    try {
      response = await this.fetcher(this.jwksURL);
    } catch (error) {
      if (error instanceof JWKSFetchError) {
        throw error;
      }
      throw new JWKSFetchError("fetch JWKS", error);
    }
    if (!response.ok) {
      throw new JWKSFetchError(`fetch JWKS: HTTP ${response.status}`);
    }

    let body: unknown;
    try {
      body = await response.json();
    } catch (error) {
      throw new JWKSFetchError("decode JWKS response", error);
    }
    if (!isJSONWebKeySet(body)) {
      throw new JWKSFetchError("decode JWKS response: keys must be an array");
    }

    let keySet: LocalJWKSet;
    try {
      keySet = createLocalJWKSet(toJoseJWKS(body));
    } catch (error) {
      throw new JWKSFetchError("create JWKS key set", error);
    }
    this.cache = { keySet, expiresAt: Date.now() + this.cacheTtlMs };
    return keySet;
  }
}

function toJoseJWKS(value: JSONWebKeySet): JoseJWKS {
  return value as unknown as JoseJWKS;
}

function isJSONWebKeySet(value: unknown): value is JSONWebKeySet {
  if (typeof value !== "object" || value === null || !("keys" in value)) {
    return false;
  }
  return Array.isArray(value.keys);
}

function claimsFromPayload(
  payload: Record<string, unknown>,
  expectedAudience: string,
): Claims {
  if (payload.iss !== DEFAULT_ISSUER) {
    throw new TokenClaimsError("invalid access token issuer");
  }
  if (typeof payload.sub !== "string" || payload.sub === "") {
    throw new TokenClaimsError("access token subject is missing");
  }

  const expiry = payload.exp;
  if (typeof expiry !== "number" || !Number.isFinite(expiry) || expiry <= Date.now() / 1000) {
    throw new TokenClaimsError("access token is expired or has no expiration");
  }

  const audience = payload.aud;
  const audienceValues = normalizeAudience(audience);
  if (audienceValues === undefined || audienceValues.length !== 1 || audienceValues[0] !== expectedAudience) {
    throw new TokenClaimsError("invalid access token audience");
  }

  const kind = payload.kind;
  if (kind !== "user" && kind !== "service") {
    throw new TokenClaimsError("invalid access token kind");
  }

  const permVer = payload.perm_ver;
  if (
    typeof permVer !== "number" ||
    !Number.isSafeInteger(permVer) ||
    permVer < 0
  ) {
    throw new TokenClaimsError("invalid access token perm_ver");
  }

  let team = "";
  if (payload.team !== undefined) {
    if (typeof payload.team !== "string") {
      throw new TokenClaimsError("invalid access token team");
    }
    team = payload.team;
  }

  return new Claims({
    subject: payload.sub,
    team,
    kind,
    permVer,
    expiry: new Date(expiry * 1000),
    audience: Array.isArray(audience) ? [...audienceValues] : audienceValues[0],
  });
}

function normalizeAudience(value: unknown): string[] | undefined {
  if (typeof value === "string") {
    return value === "" ? undefined : [value];
  }
  if (
    Array.isArray(value) &&
    value.length > 0 &&
    value.every((entry): entry is string => typeof entry === "string" && entry !== "")
  ) {
    return [...value];
  }
  return undefined;
}

function isNoMatchingKeyError(error: unknown): boolean {
  return (
    typeof error === "object" &&
    error !== null &&
    "code" in error &&
    (error as { code?: unknown }).code === "ERR_JWKS_NO_MATCHING_KEY"
  );
}
